drivers: i3c: add open-drain speed configuration API

Shreehari-AlifSemi opened the pull request:
Add new API to configure I3C open-drain speeds to meet timing requirements specified in the MIPI I3C specification for proper bus initialization and device compatibility.

Requirement (MIPI Specification for I3C Basic v1.2 (16-Dec-2024),
             Section 4.3.11.2 & Table 49):
- The MIPI I3C specification mandates different open-drain speeds during bus initialization as defined in "I3C Basic Open Drain Timing Parameters".
- The first broadcast address (7'h7E+W) must be transmitted at a slower open-drain speed to ensure visibility to all devices on the I3C bus, including legacy I2C devices. This slow speed (minimum 200ns Thigh) allows I2C devices to properly detect the I3C mode transition and disable their spike filters before switching to I3C mode. After the initial broadcast, normal I3C open-drain speeds can be used for regular operation.

Implementation:
    i3c.h
    - Add scl_od_min structure to i3c_config_controller for
      high/low period settings
    - Define Open-drain mixed bus SCL timing constants
    i3c_dw.c:
    - Add OD minimum high_ns and low_ns DT properties
    - Implement dw_i3c_init_scl_timing for runtime timing updates
    - Initialize scl_od_min configuration from device tree
    i3c_common.c:
    - Fetch current controller configuration
    - Invoke configure api inside bus_init to configure
      I3C open-drain high period for proper bus
      initialization and device compatibility.
    i3c_shell.c
    - Add SHELL support for configuring I3C open-drain speed modes.
    - Include open-drain timing configuration inside cmd_i3c_speed function.

XenuIsWatching reviewed (changes requested):
overall i think this is a necessary change as it follows that line item in the spec, but I don't think there is a need for a new api, there is an existing one for configuration. [see comment]

XenuIsWatching commented on drivers/i3c/i3c_common.c:
```diff
@@ -1380,6 +1391,15 @@ int i3c_bus_init(const struct device *dev, const struct i3c_dev_list *dev_list)
+			return ret;
+		}
+	}
+
```
can you separate these changes in to a separate commit (which will use the `i3c_configure` api from the other comment).

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1411,7 +1412,8 @@ static int i3c_dw_irq(const struct device *dev)
 }
 
-static int init_scl_timing(const struct device *dev)
+static int dw_i3c_set_speed(const struct device *dev,
```
Maybe something like this which will give flexibly on the timing allowing the application to configure min (or max) timing

```c
/**
 * @brief Configuration parameters for I3C hardware to act as controller.
 */
struct i3c_config_controller {
	/**
	 * True if the controller is to be the secondary controller
	 * of the bus. False to be the primary controller.
	 */
	bool is_secondary;

	struct {
		/** SCL frequency (in Hz) for I3C transfers. */
		uint32_t i3c;

		/** SCL frequency (in Hz) for I2C transfers. */
		uint32_t i2c;
	} scl;

	struct {
		/** SCL Open-Drain High Period in Nanoseconds */
		uint16_t high_ns;

		/** SCL Open-Drain Low Period in Nanoseconds */
		uint16_t low_ns;
	} scl_od;

	/**
	 * Bit mask of supported HDR modes (0 - 7).
	 *
	 * This can be used to enable or disable HDR mode
	 * supported by the hardware at runtime.
	 */
	uint8_t supported_hdr;
};

XenuIsWatching replied:
1. Those drivers probably should be checking `if (ctrl_cfg->scl.i3c != data->common.ctrl_config.scl.i3c)` so they do not do redundant changes... but they wouldn't be causing any issue anyway writting the same value in the registers. I can easily fix the ones that I maintain here.
2. I don't quite agree with how linux went about it there. They gave "fixed" values (when the spec only lists min and/or max values) along with a misleading name for an api `set_speed` which would confuse alot of people out there. `set_speed` would probably imply to those setting the i3c speed much like that shell command in i2c and i3c. Also, Linux does not have an existing `configure` function (and I really have no idea why...(means they have no way to handle MXDS :disappointed: )) to be considered. It also may be eventually needed they would have to be a push pull low (and likely high?) parameters for configure. This is can be an issue if that BCR[0] is 1 which will then be reported by the GETMXDS CCC which would require it's timing to be adjusted if the clk 2 data turn-around time needs to be handled differently and will need an asymmetric clock.... but I haven't come across any real-world devices (yet) that require this, but it could happen in the future.

Shreehari-AlifSemi committed:
drivers: i3c: configure open-drain speed

Add support for configuring I3C open-drain speed modes to handle
different timing requirements during bus initialization and operation.

Implementation:
- Fetch current controller configuration
- Invoke configure api inside bus_init to configure
  I3C open-drain high period for proper bus
  initialization and device compatibility.

The slow speed mode is essential for the first broadcast address
transmission to ensure visibility to all devices on the I3C bus,
particularly those in I2C mode that need to disable their spike
filters when switching to I3C mode. This requirement is specified
in "Table 49 I3C Basic Open Drain Timing Parameters" of the MIPI
ALLIANCE Specification for I3C Basic Version 1.2.

Signed-off-by: Shreehari HK <shreehari.hk@alifsemi.com>

XenuIsWatching reviewed (changes requested):
could you also update `i3c_shell` for `set_speed` adding optional parameters for setting the od low and high values

Shreehari-AlifSemi committed:
drivers: i3c: add open-drain speed support

Add SHELL support for configuring I3C open-drain speed modes.

Implementation:
- Include open-drain timing configuration inside cmd_i3c_speed function.

Signed-off-by: Shreehari HK <shreehari.hk@alifsemi.com>

Shreehari-AlifSemi commented:
> could you also update `i3c_shell` for `set_speed` adding optional parameters for setting the od low and high values

Added

Shreehari-AlifSemi commented on drivers/i3c/i3c_common.c:
```diff
@@ -1346,10 +1346,29 @@ int i3c_bus_deftgts(const struct device *dev)
 int i3c_bus_init(const struct device *dev, const struct i3c_dev_list *dev_list)
 {
+	const struct i3c_driver_api *api =
+		(const struct i3c_driver_api *)dev->api;
```
Updated

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1414,70 +1414,105 @@ static int i3c_dw_irq(const struct device *dev)
+	if (ctrl_cfg->scl_od.low_ns < config->od_tlow_min_ns) {
+		LOG_ERR("Open Drain Low Period is out of range");
+		return -EINVAL;
+	}
```
Updated

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1414,70 +1414,105 @@ static int i3c_dw_irq(const struct device *dev)
 }
 
-static int init_scl_timing(const struct device *dev)
+static int dw_i3c_init_scl_timing(const struct device *dev, void *cfg)
```
Updated the parameter type to controller configuration

Shreehari-AlifSemi commented on include/zephyr/drivers/i3c.h:
```diff
@@ -464,6 +474,14 @@ struct i3c_config_controller {
+
+		/** SCL Open Drain Low period in Nanoseconds */
+		uint32_t low_ns;
+	} scl_od;
```
Added the comments

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2124,6 +2165,15 @@ static int dw_i3c_configure(const struct device *dev, enum i3c_config_type type,
 		val = (uint32_t)(target_cfg->pid & 0xFFFFFFFF);
 		sys_write32(val, dev_config->regs + SLV_PID_VALUE);
+
+		ret = dw_i3c_init_scl_timing(dev, NULL);
```
Added a parameter validity check inside the function.

XenuIsWatching replied:
🤦 so why are you still calling it wil NULL, this will fail everytime

XenuIsWatching replied:
looking more into (and refreshing myself)... this function doesn't need to be called at all by this configure function as there is nothing to reconfigure for target mode in here. It already is called once upon it's init function

Shreehari-AlifSemi commented:
> As `od_tlow_min_ns` is now apart of the configure struct, it should loaded in to the common configuration in the driver data rather than in the config like it is now with `.common.ctrl_config.scl_od_min.low = DT_INST_PROP(n, od_tlow_min_ns),`

Removed the static initialization

XenuIsWatching commented on drivers/i3c/i3c_common.c:
```diff
@@ -1350,6 +1350,24 @@ int i3c_bus_init(const struct device *dev, const struct i3c_dev_list *dev_list)
+
+	/* Set OD timings for Slow speed */
+	ctrl_cfg.scl_od.high_ns = I3C_OD_FIRST_BC_THIGH_MIN_NS;
+	ctrl_cfg.scl_od.low_ns  = I3C_OD_TLOW_MIN_NS;
```
this should be left as is, no need to write the value here for the `high_ns`.... only the `high_ns`

but it probably should be checking if the `scl_od_min.high_ns` already meets the timing
```suggestion
	ctrl_cfg.scl_od.low_ns  = MAX(I3C_OD_TLOW_MIN_NS, ctrl_cfg.scl_od.low_ns);
```

Shreehari-AlifSemi replied:
Can you elaborate about high_ns setting here

XenuIsWatching replied:
Sorry, I meant to write "no need to write the value here for the low_ns"

XenuIsWatching commented on drivers/i3c/i3c_common.c:
```diff
@@ -1380,6 +1398,14 @@ int i3c_bus_init(const struct device *dev, const struct i3c_dev_list *dev_list)
 	}
 
+	/* Set OD timings for Normal speed */
+	ctrl_cfg.scl_od.high_ns = I3C_OD_THIGH_MAX_NS;
```
this is setting the min time to the max time?

it should just be setting it back to what it was before

Shreehari-AlifSemi replied:
Updated

XenuIsWatching commented on drivers/i3c/i3c_shell.c:
```diff
@@ -2390,8 +2396,8 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
 		      "Set I3C device speed\n"
-		      "Usage: speed <device> <speed>",
-		      cmd_i3c_speed, 3, 0),
+		      "Usage: speed <device> <speed> <od high_ns> <od low_ns>",
```
optional params need brackets around them in the help
```suggestion
		      "Usage: speed <device> <speed> [<od high_ns>] [<od low_ns>]",
```

XenuIsWatching commented on include/zephyr/drivers/i3c.h:
```diff
@@ -42,6 +42,16 @@
+#define I3C_OD_TLOW_MIN_NS           200
+#define I3C_OD_THIGH_MAX_NS          41
+#define I3C_OD_FIRST_BC_THIGH_MIN_NS 200
+#define I3C_SDR_FREQ_HZ              12500000
+#define I3C_I2C_FM_SCL_FREQ_HZ       400000
```
these 2 are unused and should be removed

Shreehari-AlifSemi replied:
Updated

XenuIsWatching commented on include/zephyr/drivers/i3c.h:
```diff
@@ -464,6 +474,22 @@ struct i3c_config_controller {
+		 * Requested minimum SCL Open-Drain LOW period in nanoseconds.
+		 */
+		uint32_t low_ns;
+	} scl_od;
```
rename so it's more clear
```suggestion
	} scl_od_min;
```

Shreehari-AlifSemi replied:
Updated

XenuIsWatching commented on include/zephyr/drivers/i3c.h:
```diff
@@ -42,6 +42,16 @@
+ *        Standard I3C SDR and I2C FM speed
+ */
+#define I3C_OD_TLOW_MIN_NS           200
+#define I3C_OD_THIGH_MAX_NS          41
```
this is only for a mixed bus so it isn't seen through the 50ns spike/glitch filter, it probably needs a better name

Shreehari-AlifSemi replied:
Updated the macro name

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2504,8 +2562,6 @@ static DEVICE_API(i3c, dw_i3c_api) = {
 		.clock_subsys = COND_CODE_1(DT_INST_PHA_HAS_CELL(n, clocks, clkid),                \
 				((clock_control_subsys_t)DT_INST_CLOCKS_CELL(n, clkid)),           \
 				((clock_control_subsys_t)0)),                                      \
-		.od_thigh_max_ns = DT_INST_PROP(n, od_thigh_max_ns),                               \
-		.od_tlow_min_ns = DT_INST_PROP(n, od_tlow_min_ns),                                 \
```
I don't think you understood my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/95747#issuecomment-3286245478
it still needs to read the property for the default values just like how the scl default is handled

Shreehari-AlifSemi replied:
> I don't think you understood my comment here: [#95747 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/95747#issuecomment-3286245478) it still needs to read the property for the default values just like how the scl default is handled

Added code to read the init configuration into scl_od_min variables from DT.

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2072,17 +2108,26 @@ static int dw_i3c_config_get(const struct device *dev, enum i3c_config_type type
+		}
+		(void)memcpy(&data->common.ctrl_config,
+			     config, sizeof(data->common.ctrl_config));
+#endif
```
you need this
```suggestion
#else
      return -ENOTSUP;
#endif
```

Shreehari-AlifSemi replied:
Updated

> you need this for when it is compiled without controller support

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,57 +1422,95 @@ static int init_scl_timing(const struct device *dev)
-
-	if (data->mode != I3C_BUS_MODE_PURE) {
+
+	if (!ctrl_cfg) {
+		LOG_ERR("Invalid Configuration");
+	}
```
this doesn't work here, it still reads it below and will hard fault..

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2385,11 +2427,27 @@ static int dw_i3c_init(const struct device *dev)
 	}
 #endif /* CONFIG_I3C_CONTROLLER */
 	dw_i3c_enable_controller(config, true);
+	enable_interrupts(dev);
+
+#ifdef CONFIG_I3C_TARGET
+	ret = dw_i3c_init_scl_timing(dev, ctrl_config);
+	if (ret != 0) {
+		LOG_ERR("Normal Speed set failed");
+		return ret;
+	}
+#endif /* CONFIG_I3C_TARGET */
 #ifdef CONFIG_I3C_CONTROLLER
+
 	if (!(ctrl_config->is_secondary)) {
 		/* Perform bus initialization - skip if no I3C devices are known. */
 		if (config->common.dev_list.num_i3c > 0) {
 			ret = i3c_bus_init(dev, &config->common.dev_list);
+		} else {
+			ret = dw_i3c_init_scl_timing(dev, ctrl_config);
+			if (ret != 0) {
+				LOG_ERR("Normal Speed set failed");
+				return ret;
+			}
```
This needs to be reverted back to what it is above as this won't work as it's not calling scl timing..

XenuIsWatching replied:
> The timing config happens twice when the device is a controller if i keep it general as before. Once by default and one more time at i3c_bus_init() function.

yes... but that's okay for it be called twice, my concern is that `i3c_bus_init()` is a public and can be called easily outside. It's also possible that `i3c_bus_init` could change in future, I'd rather see this be defensive and program it's own scl timing before it calls that `i3c_bus_init`.

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2357,13 +2406,6 @@ static int dw_i3c_init(const struct device *dev)
 	__ASSERT_NO_MSG((IS_ENABLED(CONFIG_I3C_TARGET) && ctrl_config->is_secondary) ||
 			(IS_ENABLED(CONFIG_I3C_CONTROLLER) && !ctrl_config->is_secondary));
 
-	ret = init_scl_timing(dev);
-	if (ret != 0) {
-		return ret;
-	}
-
-	enable_interrupts(dev);
-
```
This needs to be added back wtih it calling that `dw_i3c_init_scl_timing` rather than the old name

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,57 +1422,96 @@ static int init_scl_timing(const struct device *dev)
+	}
+
+	if (ctrl_cfg->scl_od_min.high_ns > I3C_OD_MIXED_BUS_THIGH_MAX_NS) {
+		if (ctrl_cfg->scl_od_min.high_ns != I3C_OD_FIRST_BC_THIGH_MIN_NS) {
+			LOG_ERR("Open Drain High Period is not in the range");
+			return -EINVAL;
+		}
```
this won't allow any other values, remove it

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,57 +1422,96 @@ static int init_scl_timing(const struct device *dev)
+
+		scl_timing = SCL_I3C_TIMING_HCNT(hcnt) | SCL_I3C_TIMING_LCNT(lcnt);
+		sys_write32(scl_timing, config->regs + SCL_I3C_OD_TIMING);
+		sys_write32(scl_timing, config->regs + SCL_I3C_PP_TIMING);
```
why is this writing to Push-Pull timing,  but it's doing no calculation or push-pull in this block

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,57 +1422,96 @@ static int init_scl_timing(const struct device *dev)
-
-	if (data->mode != I3C_BUS_MODE_PURE) {
+
+	if (!ctrl_cfg) {
+		LOG_ERR("Invalid Configuration");
+		return -EINVAL;
+	}
```
if I LOG error and return, then the user understands that it's a wrong configuration and has a chance to reconfigure properly.

Shreehari-AlifSemi replied:
Done

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1463,21 +1470,23 @@ static int init_scl_timing(const struct device *dev)
 
 	/* I2C FM+ */
-	lcnt = DIV_ROUND_UP(I3C_BUS_I2C_FMP_TLOW_MIN_NS * (uint64_t)core_rate, I3C_PERIOD_NS);
+	lcnt = DIV_ROUND_UP(I3C_BUS_I2C_FMP_TLOW_MIN_NS * (uint64_t)core_rate,
+				I3C_PERIOD_NS);
```
this should be able to fit one line as before

Shreehari-AlifSemi replied:
Done

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1463,21 +1470,23 @@ static int init_scl_timing(const struct device *dev)
 		sys_write32(BUS_I3C_MST_FREE(lcnt), config->regs + BUS_FREE_TIMING);
-		sys_write32(sys_read32(config->regs + DEVICE_CTRL) | DEV_CTRL_I2C_SLAVE_PRESENT,
-			    config->regs + DEVICE_CTRL);
+		sys_write32(sys_read32(config->regs + DEVICE_CTRL) |
+			DEV_CTRL_I2C_SLAVE_PRESENT, config->regs + DEVICE_CTRL);
```
revert this line as it's formating change

Shreehari-AlifSemi replied:
Done

XenuIsWatching commented on drivers/i3c/i3c_common.c:
```diff
@@ -1350,6 +1350,26 @@ int i3c_bus_init(const struct device *dev, const struct i3c_dev_list *dev_list)
+		return ret;
+	}
+
+	/* Set OD timings for Slow speed */
```
can this be a more descriptive comment such as reference why this needs to be done from the footnote in the i3c spec

Shreehari-AlifSemi replied:
Added nice description

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1463,21 +1470,23 @@ static int init_scl_timing(const struct device *dev)
 	/* I2C FM */
-	lcnt = DIV_ROUND_UP(I3C_BUS_I2C_FM_TLOW_MIN_NS * (uint64_t)core_rate, I3C_PERIOD_NS);
+	lcnt = DIV_ROUND_UP(I3C_BUS_I2C_FM_TLOW_MIN_NS * (uint64_t)core_rate,
+				I3C_PERIOD_NS);
```
Done

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2385,7 +2398,17 @@ static int dw_i3c_init(const struct device *dev)
 	}
 #endif /* CONFIG_I3C_CONTROLLER */
 	dw_i3c_enable_controller(config, true);
+
+	ret = dw_i3c_init_scl_timing(dev, ctrl_config);
+	if (ret != 0) {
+		LOG_ERR("Clock setting failed");
+		return ret;
+	}
+
+	enable_interrupts(dev);
+
```
revert this hunk

Shreehari-AlifSemi replied:
May I know the reason.
The purpose of keeping it here is to enable interrupt after configuring the device. It was enabling interrupt in the middle of device configuration before.

XenuIsWatching replied:
i see, i guess that's okay :shrug:

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1477,7 +1483,7 @@ static int init_scl_timing(const struct device *dev)
 		sys_write32(BUS_I3C_MST_FREE(lcnt), config->regs + BUS_FREE_TIMING);
 		sys_write32(sys_read32(config->regs + DEVICE_CTRL) | DEV_CTRL_I2C_SLAVE_PRESENT,
-			    config->regs + DEVICE_CTRL);
+				config->regs + DEVICE_CTRL);
```
revert this line

Shreehari-AlifSemi replied:
Updated

XenuIsWatching replied:
i don't see an update, did you push?

Shreehari-AlifSemi replied:
I am waiting for your response on https://github.com/zephyrproject-rtos/zephyr/pull/95747#discussion_r2357997242

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,12 +1422,23 @@ static int init_scl_timing(const struct device *dev)
+
 	/* I3C_OD */
-	hcnt = DIV_ROUND_UP(config->od_thigh_max_ns * (uint64_t)core_rate, I3C_PERIOD_NS) - 1;
+	hcnt = DIV_ROUND_UP(ctrl_cfg->scl_od_min.high_ns * (uint64_t)core_rate,
+				I3C_PERIOD_NS) - 1;
```
this should fit on one line

Shreehari-AlifSemi replied:
Updated

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,12 +1422,23 @@ static int init_scl_timing(const struct device *dev)
 	hcnt = CLAMP(hcnt, SCL_I3C_TIMING_CNT_MIN, SCL_I3C_TIMING_CNT_MAX);
 
-	lcnt = DIV_ROUND_UP(config->od_tlow_min_ns * (uint64_t)core_rate, I3C_PERIOD_NS);
+	lcnt = DIV_ROUND_UP(ctrl_cfg->scl_od_min.low_ns * (uint64_t)core_rate,
+				I3C_PERIOD_NS);
```
this should fit on one line

XenuIsWatching reviewed (commented):
can you run clang-format on these changes, that should make it easier

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2385,7 +2398,17 @@ static int dw_i3c_init(const struct device *dev)
+	enable_interrupts(dev);
+
 #ifdef CONFIG_I3C_CONTROLLER
+
```
Updated

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,12 +1421,21 @@ static int init_scl_timing(const struct device *dev)
+
+	__ASSERT((ctrl_cfg != NULL), "Controller configuration should not be NULL");
+
+	if (ctrl_cfg->scl_od_min.low_ns < I3C_OD_TLOW_MIN_NS) {
+		LOG_ERR("Open Drain Low Period is out of range");
+		return -EINVAL;
```
won't this always fail? as the first value set in the init is 24ns... which should still be valid

Shreehari-AlifSemi committed:
include: zephyr: drivers: i3c: add open-drain timing configuration

This commit adds the I3C handling of open-drain
timing configurations with the following changes:
- Add scl_od_min structure to i3c_config_controller for
  high/low period settings
- Define Open-drain mixed bus SCL timing constants

Requirement (MIPI Specification for I3C Basic v1.2 (16-Dec-2024),
             Section 4.3.11.2 & Table 49):
- The MIPI I3C specification mandates different open-drain speeds during
  bus initialization as defined in
  "I3C Basic Open Drain Timing Parameters".
- The first broadcast address (7'h7E+W) must be transmitted at a slower
  open-drain speed to ensure visibility to all devices on the I3C bus,
  including legacy I2C devices. This slow speed (minimum 200ns Thigh)
  allows I2C devices to properly detect the I3C mode transition and
  disable their spike filters before switching to I3C mode. After the
  initial broadcast, normal I3C open-drain speeds can be used for
  regular operation.

Signed-off-by: Shreehari HK <shreehari.hk@alifsemi.com>

Shreehari-AlifSemi committed:
drivers: i3c: dw: enhance open-drain timing configuration

This commit improves the I3C handling of open-drain
timing configurations with the following changes:
- Add OD minimum high_ns and low_ns DT properties
- Implement dw_i3c_init_scl_timing for runtime timing updates
- Initialize scl_od_min configuration from device tree
- Maintain backward compatibility with existing configurations

Requirement (MIPI Specification for I3C Basic v1.2 (16-Dec-2024),
             Section 4.3.11.2 & Table 49):
- The MIPI I3C specification mandates different open-drain speeds during
  bus initialization as defined in
  "I3C Basic Open Drain Timing Parameters".
- The first broadcast address (7'h7E+W) must be transmitted at a slower
  open-drain speed to ensure visibility to all devices on the I3C bus,
  including legacy I2C devices. This slow speed (minimum 200ns Thigh)
  allows I2C devices to properly detect the I3C mode transition and
  disable their spike filters before switching to I3C mode. After the
  initial broadcast, normal I3C open-drain speeds can be used for
  regular operation.

Signed-off-by: Shreehari HK <shreehari.hk@alifsemi.com>

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,12 +1421,21 @@ static int init_scl_timing(const struct device *dev)
+	if (ctrl_cfg->scl_od_min.low_ns < I3C_OD_TLOW_MIN_NS) {
+		LOG_ERR("Open Drain Low Period is out of range");
+		return -EINVAL;
+	}
```
Hmm

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -1427,12 +1421,21 @@ static int init_scl_timing(const struct device *dev)
+	__ASSERT((ctrl_cfg != NULL), "Controller configuration should not be NULL");
+
+	if (ctrl_cfg->scl_od_min.low_ns < I3C_OD_TLOW_MIN_NS) {
+		LOG_ERR("Open Drain Low Period is out of range");
```
Updated

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -2385,6 +2395,15 @@ static int dw_i3c_init(const struct device *dev)
+
+	ret = dw_i3c_init_scl_timing(dev, ctrl_config);
+	if (ret != 0) {
+		LOG_ERR("Clock setting failed");
```
Updated

Shreehari-AlifSemi commented on include/zephyr/drivers/i3c.h:
Spitted the commit

XenuIsWatching reviewed (changes requested):
sorry for all the churn in reviewing, but it looks like it's almost there now

XenuIsWatching reviewed (approved):
my only remaining concern is that this PR would change the default behavior for this driver... while the minimum of 24ns follows the spec (from the set 41ns), I'm not sure of the users of this driver. [pinging them for a review here (@gmarull @dkalowsk)].... but otherwise it looks fine to me

There were also other issues I noticed in the driver when going through the review such as the other scl ratios are not right for some of the other i2c speeds and the bus free time is following the spec for a mixed bus, but it should be setting the time for a pure bus in that case ... but that can come in another PR

ttmut commented:
Thanks for your work.

Are these additions still valid or removed in favor of the existing `configure` api?
> 
> * Add set_speed callback to i3c_driver_api structure
> * Implement set speed api in DW driver
